config/dynamic.c

```diff
@@ -16,11 +16,546 @@
  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
  */
 
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <sys/un.h>
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <sys/select.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <rte_log.h>
+#include <rte_lcore.h>
+#include <rte_cycles.h>
+#include <rte_malloc.h>
+
+#include "gatekeeper_net.h"
+#include "gatekeeper_main.h"
 #include "gatekeeper_config.h"
+#include "gatekeeper_launch.h"
+
+/* TODO Get the install path from the configuration file. */
+#define LUA_DY_BASE_DIR     "./lua"
+#define DYNAMIC_CONFIG_FILE "dylib.lua"
+
+/*
+ * The cast "(uint16_t)" is needed because of
+ * the strict compilation check of DPDK,
+ * without uint16_t, it pops an error message
+ * "error: large integer implicitly truncated to
+ * unsigned type [-Werror=overflow]."
+ */
+static const uint16_t MSG_MAX_LEN = (uint16_t)~0U;
+
+static struct dynamic_config config;
+
+/*
+ * Return values:
+ * 	0: Write @nbytes successfully.
+ * 	-1: Connection closed by the client.
+ */
+static int
+write_nbytes(int conn_fd, const char *msg_buff, int nbytes)
+{
+	int send_size;
+	int tot_size = 0;
+
+	if (nbytes == 0)
+		return 0;
+
+	while ((send_size = write(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += send_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the client is closed. */
+	if (send_size == 0) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"dyn_cfg: Client disconnected\n");
+		return -1;
+	}
+
+	if (send_size < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to write data to the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+reply_client_message(int conn_fd, const char *reply_msg, uint16_t reply_len)
+{
+	int ret;
+	uint16_t nlen = htons(reply_len);
+
+	/* The first two bytes: the length of the message in network order. */
+	ret = write_nbytes(conn_fd, (char *)&nlen, sizeof(nlen));
+	if (ret != 0)
+		return -1;
+
+	/* Sending the message. */
+	ret = write_nbytes(conn_fd, reply_msg, reply_len);
+	if (ret != 0)
+		return -1;
+
+	return 0;
+}
+
+static int
+process_client_message(int conn_fd,
+	const char *msg, int msg_len, lua_State *lua_state)
+{
+	/*
+	 * TODO Implement the functionalities to process the clients' request.
+	 *
+	 * Gatekeeper: Adding, listing, and removing BP components;
+	 *
+	 * Gatekeeper: Adding, listing, removing IP ranges
+	 * handled by the global LPM table in GKs;
+	 *
+	 * Gatekeeper and Grantor: Listing the ARP and ND table.
+	 * This is important for network diagnosis.
+	 *
+	 * Grantor: Updating the policy GTs run.
+	 */
+
+	int ret;
+	size_t reply_len;
+	const char *reply_msg;
+	const char *CLIENT_PROC_ERROR =
+		"function reply() return -1, \"Dynamic Configuration cannot process the request. end\"\n";
+
+	if (msg_len == 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: the received message is an empty string\n"); 
+		return -1;
```

Comment:
1. Make `RTE_LOG()` a warning and change type to `LUA`.

2. Replace `return -1` with `return reply_client_message(conn_fd, CLIENT_EMPTY_ERROR, strlen(CLIENT_EMPTY_ERROR));` There's no reason to abort the connection since we can gracefully handle this case.

You can define `CLIENT_EMPTY_ERROR` as follows:
```
const char *CLIENT_EMPTY_ERROR =
"Dynamic configuration cannot process the request: the request is empty.";
```